📖 Small improvements to migration instructions (#12328)
kubernetes-sigs/cluster-api: docs/book/src/developer/providers/migrations/v1.10-to-v1.11.md

  Please refer to [API Changes](#api-changes) for more details about the changes introduced by this release.

  Note: it is technically possible for providers to keep using v1beta1 types from CAPI v1.11, but this is 
  not recommended because it will lead to additional conversion calls.

  Additionally, in v1.11 all the CAPI utils like e.g. `IsControlPlaneMachine` are now using v1beta2 API types. 
  This might lead to additional work for providers to keep using v1beta1 types from CAPI v1.11 (you have to fork

>   all the all utils the providers is using from an older CAPI release or replace them with something else).

Reviewer: ```suggestion
  all utils the provider is using from an older CAPI release or replace them with something else).
```

```go
import (
	...

>     v1beta1conditions "sigs.k8s.io/cluster-api/util/deprecated/v1beta1/conditions"
>     v1beta2conditions "sigs.k8s.io/cluster-api/util/deprecated/v1beta1/conditions/v1beta2"

Reviewer 2: In CAPV we aliased them the other way around:
```suggestion
    deprecatedconditions "sigs.k8s.io/cluster-api/util/deprecated/v1beta1/conditions"
    deprecatedv1beta2conditions "sigs.k8s.io/cluster-api/util/deprecated/v1beta1/conditions/v1beta2"
```
So we e.g. don't messup:
* this `v1beta1conditions "sigs.k8s.io/cluster-api/util/deprecated/v1beta1/conditions"`
* vs `v1beta1conditions "sigs.k8s.io/cluster-api/util/conditions/v1beta1"`

import (
	...
    "sigs.k8s.io/cluster-api/util/conditions"
    deprecatedv1beta1conditions "sigs.k8s.io/cluster-api/util/conditions/deprecated/v1beta1"
    "sigs.k8s.io/cluster-api/util/patch"
)
```

- the `conditions` package provides access to utils for managing `metav1.Conditions` in `status.conditions`
- the `deprecatedv1beta1conditions` package alias provides access to utils for managing `clusterv1.Conditions` in `status.deprecated.v1beta1.conditions`
- the `patch` package provides access to utils for patching objects in this phase

Important!

> - Please pay special attention to use `sigs.k8s.io/cluster-api/util/patch` import everywhere,
>   because using `sigs.k8s.io/cluster-api/util/deprecated/v1beta1/patch` by mistake could in some cases lead to dropping conditions at runtime.

Reviewer: ```suggestion
```
In this stage we want folks to use util/patch

Author: Yes, we want folks to use util/patch
This sentence is about making sure users do not forget to change in all the places what they were using in stage1

I tried to clarify, PTAL

Reviewer: Thx. Yeah I was misreading this sentence, my bad

import (
	...
    "sigs.k8s.io/cluster-api/util/conditions"
    deprecatedv1beta1conditions "sigs.k8s.io/cluster-api/util/conditions/deprecated/v1beta1"
    "sigs.k8s.io/cluster-api/util/patch"
)
```

- the `conditions` package provides access to utils for managing `metav1.Conditions` in `status.conditions`
- the `deprecatedv1beta1conditions` package alias provides access to utils for managing `clusterv1beta2.Conditions` in `status.deprecated.v1beta1.conditions`
- the `patch` package provides access to utils for patching objects in this phase

Important!
- Please pay special attention to use `sigs.k8s.io/cluster-api/util/patch` import everywhere,

>   because if by mistake you use `sigs.k8s.io/cluster-api/util/deprecated/v1beta1/patch` in some files this won't lead to compile errors
>   but conditions are going to be dropped at runtime.

Reviewer 2: ```suggestion
  because using `sigs.k8s.io/cluster-api/util/deprecated/v1beta1/patch` by mistake could in some cases lead to dropping conditions at runtime,
  while not having compile errors.
```

- the `conditions` package provides access to utils for managing `metav1.Conditions` in `status.conditions`
- the `patch` package provides access to utils for patching objects in this phase

Important!
- Please pay special attention to use `sigs.k8s.io/cluster-api/util/patch` import everywhere,

>   because using `sigs.k8s.io/cluster-api/util/deprecated/v1beta1/patch` by mistake could in some cases lead to dropping conditions at runtime.

Reviewer: Same here

## Annex

### Imports for conditions and patch helper utils

In order to help users to transition away from the CAPI conditions type, CAPI v1.11 supports
different versions of conditions and patch helper utils.

> Following table should help in using the right version

Reviewer: ```suggestion
Following table should help to pick the right utils.
```

- the `conditions` package provides access to utils for managing `metav1.Conditions` in `status.conditions`
- the `patch` package provides access to utils for patching objects in this phase

Important!
- Please pay special attention to use `sigs.k8s.io/cluster-api/util/patch` import everywhere,

>   because if by mistake you use `sigs.k8s.io/cluster-api/util/deprecated/v1beta1/patch` in some files this won't lead to compile errors
>   but conditions are going to be dropped at runtime.

Reviewer 2: ```suggestion
  because using `sigs.k8s.io/cluster-api/util/deprecated/v1beta1/patch` by mistake could in some cases lead to dropping conditions at runtime,
  while not having compile errors.
```

## Annex

### Imports for conditions and patch helper utils

In order to help users transition away from CAPI conditions type, CAPI v1.11 supports
different versions of conditions and patch helper utils.

Following table should help in using the right version

| Field to change                                                       | Import for condition util                                                                | Import for patch helper                                              |
|-----------------------------------------------------------------------|------------------------------------------------------------------------------------------|----------------------------------------------------------------------|

> | `status.conditions` of type `clusterv1beta1.Conditions`               | v1beta1conditions "sigs.k8s.io/cluster-api/util/deprecated/v1beta1/conditions"           | v1beta1patch "sigs.k8s.io/cluster-api/util/deprecated/v1beta1/patch" |
> | `status.v1beta2.conditions` of type `[]metav1.Conditions`             | v1beta2conditions "sigs.k8s.io/cluster-api/util/deprecated/v1beta1/conditions/v1beta2"   | v1beta1patch "sigs.k8s.io/cluster-api/util/deprecated/v1beta1/patch" |
> | `status.deprecated.v1beta1.conditions` of type `clusterv1.Conditions` | deprecatedv1beta1conditions "sigs.k8s.io/cluster-api/util/conditions/deprecated/v1beta1" | "sigs.k8s.io/cluster-api/util/patch"                                 |
> | `status.conditions` of type `[]metav1.Conditions`                     | "sigs.k8s.io/cluster-api/util/conditions"                                                | "sigs.k8s.io/cluster-api/util/patch"                                 |

Reviewer 2: ```suggestion
| `status.conditions` of type `clusterv1beta1.Conditions`               | `v1beta1conditions "sigs.k8s.io/cluster-api/util/deprecated/v1beta1/conditions"`           | `v1beta1patch "sigs.k8s.io/cluster-api/util/deprecated/v1beta1/patch"` |
| `status.v1beta2.conditions` of type `[]metav1.Conditions`             | `v1beta2conditions "sigs.k8s.io/cluster-api/util/deprecated/v1beta1/conditions/v1beta2"`   | `v1beta1patch "sigs.k8s.io/cluster-api/util/deprecated/v1beta1/patch"` |
| `status.deprecated.v1beta1.conditions` of type `clusterv1.Conditions` | `deprecatedv1beta1conditions "sigs.k8s.io/cluster-api/util/conditions/deprecated/v1beta1"` | `"sigs.k8s.io/cluster-api/util/patch"`                                 |
| `status.conditions` of type `[]metav1.Conditions`                     | `"sigs.k8s.io/cluster-api/util/conditions"`                                                | `"sigs.k8s.io/cluster-api/util/patch"`                                 |
```

Important!

> - Please pay special attention to use the correct patch helper import everywhere, because if by mistake you use a wrong
>   onw in some files this won't lead to compile errors but conditions are going to be dropped at runtime.

Reviewer 2: ```suggestion
- Please pay special attention to use the correct patch helper import everywhere, because using a wrong
  one could in some cases lead to dropping conditions at runtime while not having compile errors.  

```
